### DISCUSSION TITLE
Inventory icon for 3D items

### DISCUSSION THREAD

**Discussion by YellowZaki** [Kas-tle/java2bedrock.sh#718] [Feature Requests]
In Minecraft java, 3D items automatically get their inventory icon from their 3D model. 
I would like to know if this is posible in bedrock.

**Comment by typicalsmc** [comment]
Yes

**Comment by Kas-tle** [comment]
Sort of but it's not so simple. One can use blocks, but there are some significant downsides. For one thing, blocks rendering in the inventory is extremely glitchy. This is a long-standing bug that Mojang has shown little interest in fixing: https://bugs.mojang.com/browse/MCPE-123305.

Secondly, blocks have an extremely limited set of components compared to items. This means things like tool break speeds are more or less a no go. These items would be useful for nothing other than their cosmetic utility. 

Finally, even if the rendering issues were fixed, there is still little control over the appearance of the icon itself. The position that the model is rendered from is fixed and there is no way to change it, beyond rotating the entire model. Because of this, I believe custom sprites are the best option for most use cases. Geyser does not currently support blocks, so this is also the only solution for the time being.

**Reply by typicalsmc** [reply]
> Sort of but it's not so simple. One can use blocks, but there are some significant downsides. For one thing, blocks rendering in the inventory is extremely glitchy. This is a long-standing bug that Mojang has shown little interest in fixing: https://bugs.mojang.com/browse/MCPE-123305.
> 
> Secondly, blocks have an extremely limited set of components compared to items. This means things like tool break speeds are more or less a no go. These items would be useful for nothing other than their cosmetic utility.
> 
> Finally, even if the rendering issues were fixed, there is still little control over the appearance of the icon itself. The position that the model is rendered from is fixed and there is no way to change it, beyond rotating the entire model. Because of this, I believe custom sprites are the best option for most use cases. Geyser does not currently support blocks, so this is also the only solution for the time being.

We could make a guide on how to export the item 2d sprite in Blockbench btw, it is kind of easy, or a js script to do it automatically would be nice

**Comment by YellowZaki** [comment]
It would be nice to be able to set the inventory item texture to a sprite you want. I guess this can only be done in geyser side, not in this tool.

I've been looking if this is possible in java (like trident is 2D item in inventory and 3D in hand) but unfortunately it is not possible to do it for custom items.

**Comment by YellowZaki** [comment]
I have came up with an idea to make adding 2D icons for 3D models easier.
Let's say you have a custom model named rubypickaxe.json which is a 3D model. If you want to show a 2D icon, you have to:
1. Add the 2D icon named rubypickaxe.png to the folder: textures/geyser/geyser_custom
2. Add an entry to the item_texture.json: 
```
    "rubypickaxe": {
      "textures": "textures/geyser/geyser_custom/rubypickaxe"
    },
```
3. Set `"icon": "rubypickaxe"` in geyser_mappings.json


This process is tedious and could be automated by passing a new argument (a directory or a zip containing all the 2D icon pngs) to the converter.sh.

**Reply by Kas-tle** [reply]
I have now implemented this. See the updated Readme for info on how to use it. 
https://github.com/Kas-tle/java2bedrock.sh#local

Models will also now have persistent identifiers across conversions provided the item and predicate values remain the same.

**Reply by YellowZaki** [reply]
> I have now implemented this. See the updated Readme for info on how to use it. https://github.com/Kas-tle/java2bedrock.sh#local
> 
> Models will also now have persistent identifiers across conversions provided the item and predicate values remain the same.

Thank you. But I think it could be done easier. I mean, with that way, we need to make the sprites.json, add the predicates, create a new Bedrock resource pack containing the specificed sprite textures...

It could be done by just passing a directory/zip containing containing all the sprites .png. The name of the .png should match the model. It is much easier.

```
for each spriteName.png in spriteDir:
     if there is a model matching spriteName in the bedrock pack:
         copy the sprite.png to textures/geyser/geyser_custom directory
         add the entry to item_texture.json
         add the spriteName to "icon" from geyser_mappings.json
```

**Reply by Kas-tle** [reply]
You can't just "match the name" as people might use duplicate names. You would have to match the whole path of the model. This would also require people to upload the same image twice if they elect to use the same icon. Therefore, I will not be implementing this.

**Reply by RedxAx** [reply]
i have an idea, but it will slow down the servers, and it might be hard (:
it should render the java model  icon, and then make it into a bedrock 2D icon!,
hard but good, and to not slow down the servers, make this only avalable for the localy runned script.
